### DISCUSSION TITLE
ADC read locked forever with CONFIG_DEBUG_OPTIMISATION=y on STM32U5

### DISCUSSION THREAD

**Discussion by tpennors** [zephyrproject-rtos/zephyr#52261] [Q&A]
**Describe the bug**
ADC read locked forever with `CONFIG_DEBUG_OPTIMISATION=y` on STM32U5
I used gdb to analyse this. I saw that the code gets locked in : 
`drivers/adc/adc_context.h: k_sem_take(&ctx->sync, K_FOREVER);`
I checked and `adc_stm32_isr` is never called.

When using `CONFIG_NO_OPTIMISATION=y`, everything is fine.

**Temporary fix**
I found a temporary fix but I would like to understand if it is a real bug from the driver. My solution is to add a very little waiting period between those two calls: `LL_ADC_EnableIT_EOC(adc);` and `adc_context_start_read(&data->ctx, sequence);`

**Other**
I didn't use overlay as it is described in `sample/drivers/adc/boards/stm32l562e_dk.overlay`, would that be a prolem ?

**Comment by nordicjm** [comment]
Probably a stack overflow, increase the stack sizes and try again

**Comment by tpennors** [comment]
Hello,

I tried to increase stack size and still doesn't work. Hardware interrupt is never received. I would have think of a timing between writing to registers which is not the same between optimizations. Did anyone see this kind of behaviour ?

**Comment by FRASTM** [comment]
When building the samples for nucleo_u575zi_q    with **CONFIG_DEBUG_OPTIMIZATIONS=y**
The measurement is correct on ADC1 channel 1 (available on Arduino connector CN9 pin A5)
Of course is the overlay required to test  (try to connect pin A5 to 3.3v line)

```
ADC reading:                                                                                                                   
- adc@42028000, channel 1: 16383 = 3299 mV   
```

**Comment by legoabram** [comment]
@tpennors We might be running into a similar problem in Zephyr v3.5.0-rc3. Did you ever find a solid answer to this problem?
